cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,218 @@
+    operator()(column_view const& in,
+               double quantile,
+               interpolation interp,
+               bool is_sorted,
+               cudf::order order,
+               cudf::null_order null_order,
+               rmm::mr::device_memory_resource *mr =
+               rmm::mr::get_default_resource(),
+               cudaStream_t stream = 0)
+    {
+        CUDF_FAIL("non-arithmetic types are unsupported");
+    }
+
+    template<typename T>
+    typename std::enable_if_t<std::is_arithmetic<T>::value, std::unique_ptr<scalar>>
+    operator()(column_view const& in,
+               double quantile,
+               interpolation interp,
+               bool is_sorted,
+               cudf::order order,
+               cudf::null_order null_order,
+               rmm::mr::device_memory_resource *mr =
+                 rmm::mr::get_default_resource(),
+               cudaStream_t stream = 0)
+    {
+        auto null_offset = null_order == cudf::null_order::AFTER ? 0 : in.null_count();
+        double result{};
+
+        if (not is_sorted)
+        {
+            table_view const in_table { { in } };
+            auto in_sortmap = sorted_order(in_table, { order }, { null_order });
+            auto in_sortmap_begin = in_sortmap->view().begin<size_type>();
+            auto in_begin = in.begin<T>() + null_offset;
+
+            auto source = [&](size_type location) {
+                auto idx = detail::get_array_value<size_type>(in_sortmap_begin, location);
+                return detail::get_array_value<T>(in_begin, idx);
+            };
+
+            result = select_quantile<double>(source,
+                                             in.size() - in.null_count(),
+                                             quantile,
+                                             interp);
+        } else {
+            auto in_begin = in.begin<T>() + null_offset;
+            auto source = [&](size_type location) {
+                return detail::get_array_value<T>(in_begin, location);
+            };
+
+            result = select_quantile<double>(source,
+                                             in.size() - in.null_count(),
+                                             quantile,
+                                             interp);
+        }
+
+        return std::make_unique<numeric_scalar<ScalarResult>>(result);
+    }
+};
+
+} // anonymous namespace
+
+std::unique_ptr<scalar>
+quantile(column_view const& in,
+         double quantile,
+         interpolation interp,
+         bool is_sorted,
+         cudf::order order,
+         cudf::null_order null_order)
+{
+        if (in.size() == in.null_count()) {
+            return std::make_unique<numeric_scalar<ScalarResult>>(0, false);
+        }
+
+        if (in.size() == 1) {
+            return type_dispatcher(in.type(), detail::pick_functor{}, in, 0);
+        }
+
+        // if (not is_sorted) {
```

Comment:
Again, compilation time increases by ~5m when uncommented.

Reply:
Why is this necessary anyway? I would think that clamping `quantile` between 0.0 and 1.0 would be sufficient. This is what `quantile_index` does.

Reply:
Necessary? No. It's only a space optimization. Since we only need the min/max element id, we don't have to allocate memory for all of the sorted indices if we use `min_element`/`max_element`.

Reply:
I'd prefer to leave it out, since rmm makes memory allocation fast, and leaving it in makes compilation incredibly slow. It was in the legacy implementation, so I added it here. Ok if I get rid of it?